### ISSUE TITLE
LRScheduler's batch_idx_ includes validation batches

### ISSUE THREAD

**Issue by thomasjpfan** [skorch-dev/skorch#418]
`LRScheduler`'s `batch_idx_` counts both validation and training batches. This means the learning rate is updated `CyclicLR` during validation batch steps.

**Comment by BenjaminBossan** [comment]
> I would like to use the `train_loss` or `train_batch_size` in the history as a proxy

Let's try to prevent using proxies, since this creates unnecessary coupling. Would it be enough to look at the `training` argument passed to `on_batch_end` ([here](https://github.com/dnouri/skorch/blob/5542209d7d3d1b30bc6bfb87e11b0278fb27f89e/skorch/callbacks/lr_scheduler.py#L148)) and only increment if `training is True`?

**Comment by thomasjpfan** [comment]
Using `training` will work as long as the learning rate scheduler records the number of training batches in the history. This way the total number of training batches can be calculated when training resumes.

**Comment by BenjaminBossan** [comment]
> Using `training` will work as long as the learning rate scheduler records the number of training batches in the history

You mean it must be in the history in case that after loading the model only, you would like to continue training or are there other cases that I overlook? In that case, should we rather have the net itself count the number of training batches?

**Comment by thomasjpfan** [comment]
> in case that after loading the model only, you would like to continue training

This is the use case I am considering.

> should we rather have the net itself count the number of training batches?

That works too. We can place the counting logic  in `NeuralNet.on_batch_end`.

**Comment by ottonemo** [comment]
If I understand correctly, the current proposal is to have two attributes in `NeuralNet`:
1. `batch_count_train_`
2. `batch_count_valid_`
both holding the current count of processed batches. The upper bound of this index is not known in advance (since there might be a dynamic amount of batches per epoch) and we can't guarantee that this number is evenly divisible by a fixed batch size (same argument). The last points might get in the way when users expect this to hold (i.e. `num_epochs = (batch_count_train + batch_count_valid)/batch_size` or something similar).

The purpose of such a counter is to have a global step count. This does not solve the problem of having no *local* step count, i.e. there is no way to know how many training batches currently have passed without coupling to `train_loss` in the history (`batch_count = len(history[-1, 'batches', :, 'train_loss'])`).
We could, of course, introduce 4 variables, one for global and one for local count or track all local counts in a list but this is just reinventing the wheel (history), I guess.

Maybe we provide something like this:
```python
global_train_steps = sum(history[:, 'batches', :, 'is_training'])
local_train_steps = sum(history[-1, 'batches', :, 'is_training'])
```
for which we would need to introduce a new history key (`is_training`). What I dislike about this is that it feels incredibly wasteful (`num_epochs * num_batches * m` byte to count 2 numbers).

So in the end I think that introducing storage to `NeuralNet` is preferable. We could think about adding these counts to `history` instead (e.g. as attributes, `history.current_train_steps`).

**Comment by thomasjpfan** [comment]
I would prefer this information to be in the history. This would keep track of state when the history gets loaded from a file.

What do you think of adding `history[:, 'batch_count_train']` and `history[:, 'batch_count_valid']` to keep track of the count?

**Comment by BenjaminBossan** [comment]
>     2. The last points might get in the way when users expect this to hold (i.e. `num_epochs = (batch_count_train + batch_count_valid)/batch_size` or something similar).

Do you have an example in mind where someone would make that assumption? Would it be enough to document the behavior somewhere ("If you want to know the number of training batches, use this key from history...")

> I would prefer this information to be in the history.

History seems to be the right place for this kind of information.

> What do you think of adding `history[:, 'batch_count_train']` and `history[:, 'batch_count_valid']` to keep track of the count?

This looks reasonable. That number would be the non-cumulative number of batches and if you need the total, you should sum them. In theory, the number of batches can be misleading when not taking account of `'train_batch_size'` and `'valid_batch_size'`, but for most practical purposes, that might not be too important.
